### DISCUSSION TITLE
How to make “as greedy as possible” client? Ow how to make high-speed streaming?

### DISCUSSION THREAD

**Discussion by AskAlexSharov** [anacrolix/torrent#741] [Q&A] [answered]
Which download everything as fast as it can

**Comment by AskAlexSharov** [comment]
I found something https://github.com/anacrolix/torrent/issues/150#issuecomment-289264973

**Comment by anacrolix** [comment]
The current way to do this would be set a very large MaxUnverifiedBytes for the Client and Torrent (experiment with this for performance reasons), and call Torrent.DownloadAll. This should cause the client to download as much as it can from all connections without overlap until endgame mode.

**Comment by AskAlexSharov** [comment]
but MaxUnverifiedBytes=MaxInt - behave same as default value MaxUnverifiedBytes=0 - means unlimited.
no?

**Reply by anacrolix** [reply]
Yes you are right. I forgot that unlimited was the default. There was an issue recently where someone had performance issues with that for very large torrents.

**Comment by AskAlexSharov** [comment]
Just FYI: we have 200gb of torrents (80 small files, and 20 large). We downloading by internet (not by private network). For now - most greedy way we found is "download files 1-by-1" sequentially. (we didn't experiment yet with `io.Copy(io.Discard, t.NewReader())` and with `MaxUnverifiedBytes`).

**Comment by anacrolix** [comment]
You'll have to clarify what your goal is: Do you want to get all the data as quickly as possible with no care for what order it arrives in, or do you want to stream the entire torrent, file-by-file? Could you maybe explain what you're trying to do, those use-cases are quite different.

**Comment by AskAlexSharov** [comment]
Goal: 
- Ethereum came back to you :-) We working on alternative to go-ethereum - project https://github.com/ledgerwatch/erigon 
- Using Bittorrent to sync historical data - not new hot data, but old and immutable.
- Main goal: download everything as fast as possible (Seeders have upload ratelimiter) - and just start using files locally (and seed them). 200Gb or 500Gb - depends on network: ~80 files < 1gb, ~20 files > 5gb, maximum 40gb file.
- And use some seedboxes providers to provide cheap data availability. 
- I believe streaming feature - may be great for us: to reduce amount of random read/writes - and make us more slow-disk-friendly. Because we doing some heavy data processing and seed torrents in same time - better reduce impact of torrent to disk. I already see very low RAM usage in anacrolix/torrent - great. (having less goroutines is also helpful). 

Problem: 
- When we download all torrents in same time - we see download rate 2Mb/s (probably because of too much concurrency and requests prioritization). When download 1-by-1 - can see >50Mb/s (depends on seeders availability). I think it's not related to torrent size, but related to amount of torrents. Can provide some pprof picture if need.
- Other observation on seedbox.io (they using rTorrent): if I have 2 seedboxes - they almost never see each other - because there is limit on peers amount (and no way to configure it). They seeding to geo-locally far peers with 300kb/s. But if stop all torrents and start 1: seedboxes see each other and can sync 100mb/s. So, amount of torrent files may lead to weird results: some files will never sync, seed speed is very low, all reads/writes are random, etc... Strategy "download 1-by-1" may also help in this case.

**Reply by anacrolix** [reply]
> Goal:
> 
> * Ethereum came back to you :-) We working on alternative to go-ethereum - project https://github.com/ledgerwatch/erigon
> * Using Bittorrent to sync historical data - not new hot data, but old and immutable.
> * Main goal: download everything as fast as possible (Seeders have upload ratelimiter) - and just start using files locally (and seed them). 200Gb or 500Gb - depends on network: ~80 files < 1gb, ~20 files > 5gb, maximum 40gb file.

If you want to seed files you've obtained, but also use the data in its file-orientation, and you have sufficient disk space, then a file-based storage implementation would probably suit best, I expect you've already done this (it's also the default). Definitely downloading as fast as possible means minimizing overlap and having the flattest possible prioritization structure. If you want some files to finish sooner, or complete whole files before moving on, that would be a slightly different case. Similarly if you're storing on something like S3 or whatever, you can seed from that, but you will want to stream or batch the data from local storage.

> * And use some seedboxes providers to provide cheap data availability.
> * I believe streaming feature - may be great for us: to reduce amount of random read/writes - and make us more slow-disk-friendly. Because we doing some heavy data processing and seed torrents in same time - better reduce impact of torrent to disk. I already see very low RAM usage in anacrolix/torrent - great. (having less goroutines is also helpful).

This might be a bit different: You would want to use a memory-based storage implementation, and then stream out the data to truly mazimize disk write speed. That also makes seeding difficult though, since you'd want to seed from the files and probably not from memory as you wouldn't want to keep everything in memory. There's no write cache in anacrolix/torrent yet, that would help with ordering writes to disk. Also you would make use of a smaller "unverified bytes" window in that case, you wouldn't want to readahead more than the size of your cache so as to ensure flushes were able to write sequentially.

> 
> Problem:
> 
> * When we download all torrents in same time - we see download rate 2Mb/s (probably because of too much concurrency and requests prioritization). When download 1-by-1 - can see >50Mb/s (depends on seeders availability). I think it's not related to torrent size, but related to amount of torrents. Can provide some pprof picture if need.

More information helps! Please send it through. Have you checked CPU utilization? anacrolix/torrent should have some CPU issues with very large torrents and an unbounded max verified bytes (it will consider the entire space of possible requests to optimize requests to each peer. In the case where you're not streaming, and aren't constraining the options, the space can be huge.

> * Other observation on seedbox.io (they using rTorrent): if I have 2 seedboxes - they almost never see each other - because there is limit on peers amount (and no way to configure it). They seeding to geo-locally far peers with 300kb/s. But if stop all torrents and start 1: seedboxes see each other and can sync 100mb/s. So, amount of torrent files may lead to weird results: some files will never sync, seed speed is very low, all reads/writes are random, etc... Strategy "download 1-by-1" may also help in this case.

I'm not quite sure I follow this, but you can up the number of peers allowed per Torrent. anacrolix/torrent hasn't been heavily optimized for seeding as I've not operated it in that situation in a performance critical situation. It should upload fine, but I can't make any claims as to how fast. It's possible that if you're finding uploads to distance peers to be slow, there is plenty of room for optimization.

**Reply by AskAlexSharov** [reply]
> memory-based storage implementation 

We can't: now in worst case on 1 machine have 400Gb of immutable torrents and also project with 7Tb database. And DB has some hot data. 

> file-based storage 

Using mmap now, just because our db is also mmap and we have an experience. (seems it open/close file on every read? or i miss-understand the code?)

> More information helps! 

I will try to collect tomorrow. As I remember it used 140% on 16-core machine, but more details tomorrow.

> In the case where you're not streaming, and aren't constraining the options, the space can be huge.

Yes, will try. FYI: I did just try qBittorrent client - and it has checkbox "download in sequential order" - if check it download 3 files at a time.

> finding uploads to distance peers to be slow

true, on neighbour servers it show 100mb/s now. Our network is not very big now (~50 fully downloaded servers around the world - just some users of our project). 

Great reply - I need get sleep with it.

**Reply by anacrolix** [reply]
Yeah there's nothing special about sequentially downloading files when you intend to download the entire torrent anyway. I expect that you're just finding it faster because the CPU load is lower if you have less pieces prioritized for download at once.

**Comment by AskAlexSharov** [comment]
Some pprof wend do .DownloadAll() on list of files (200gb of total size). 
It shows now 10Mb/s (i'm not sure why not 2mb/s as I saw before).

![cpu](https://user-images.githubusercontent.com/46885206/166202279-36bd1fe6-6525-4cb6-af4a-ba3c60df9b67.png)

**Reply by anacrolix** [reply]
Thanks that's very helpful! Switching to a generic heap implementation is critical, and there are some other optimizations to be made in the request heaping. Finding a generic heap implementation would be the lowest hanging fruit, let me know if you can help out.

**Reply by AskAlexSharov** [reply]
FYI: something also happening here https://www.scylladb.com/2022/04/27/shaving-40-off-googles-b-tree-implementation-with-go-generics/

**Reply by anacrolix** [reply]
Yeah https://github.com/ajwerner/btree is currently used for piece prioritization, I need something similar for the requests. Using a generic btree would probably be faster than the heap in use to be honest.

**Reply by AskAlexSharov** [reply]
I dived into heap.Init place and found next:
- I have PieceSize=2mb and I thought: 1. it will reduce amount of network calls 2. reduce amount of random-reads on server 3. reduce amount of hashes 
- But seems there one more variable TorrentSpec.ChunkSize and it's 16KiB.

It means: 
- PieceSize only reduce amount of hashes 
- ChunkSize must be used to - request more sequential data per 1 call and to reduce amount of network call

So, looks like ChunkSize = 4 * defaultChunkSize - greatly reduced heap size.

ChunkSize: 16 * 1024 -> heap size 1.1M, 4 * 16 * 1024 -> heap size 300K, 16 * 16 * 1024 -> heap size 70K

But I don't know - how much I can safely increase ChunkSize? What it depends on?

**Reply by AskAlexSharov** [reply]
I seems touched hardcoded: `MaxLength: 256 * 1024`

**Reply by anacrolix** [reply]
Yeah increasing chunksize is one option. I think it could increase congestion for other outbound messages, but that's only a guess. I think switching to a generic heap implementation, or something else generic is the first place to start.

**Reply by AskAlexSharov** [reply]
PR to Make pp.Decoder.MaxLength configurable: https://github.com/anacrolix/torrent/pull/743

**Reply by AskAlexSharov** [reply]
In my head "network congestion" is the problem which must be solved by TCP, not by APP. 
Or do you have some specific meaning of "network congestion"?

**Reply by anacrolix** [reply]
Yeah I mean congestion in that large messages can starve out control messages, as the BitTorrent wire protocol is streamed. It's pretty minor, it would only be an issue for very slow connections. Please also try the current `anacrolix` branch, I have a generic heap implementation in use from https://github.com/anacrolix/torrent/commit/05a3cd9697845261118e6265fac69724d8a81939.

**Reply by AskAlexSharov** [reply]
anacrolix branch cpu/alloc_objects:
![cpu](https://user-images.githubusercontent.com/46885206/167238192-a706ef86-4d51-444a-8ab0-6daab5c6cfee.png)
![mem](https://user-images.githubusercontent.com/46885206/167238226-7c73b5bb-c732-4341-9880-70f727ffa9c8.png)

**Reply by anacrolix** [reply]
Thank you so much for getting these profiles. Some observations: CPU usage for request ordering remains the same. The typed roaring bitmaps look like they're causing allocations in the piece ordering. I don't know if that's new, I would need an allocation profile for the code before the change (I may be able to infer it from running it myself elsewhere and just checking if there are allocations). The newer code does look to be using about 1/2 the CPU (150% -> 75% in the provided profiles). There looks like there's plenty of optimization to be made for the request ordering less function, I have some ideas.

**Reply by anacrolix** [reply]
The allocations are halved back to their original state in https://github.com/anacrolix/torrent/commit/9a9c7dee007fd6b10e4a5e034ce211a1291638e5, and reduced more than originally in https://github.com/anacrolix/torrent/commit/2b1f539aa27bdc047214b7dfff10234d69a5ea93. See also https://github.com/RoaringBitmap/roaring/issues/94. This seems to provide a small CPU improvement.

The request ordering CPU improvements are pending.

**Reply by anacrolix** [reply]
It's all in. Please try the branch `reuse-iters`. I've applied extremely aggressive optimizations there. In my testing for regular downloading, the request calculations are roughly ~20x faster or so. They're now significantly less than the cost of decoding the incoming data.

**Reply by AskAlexSharov** [reply]
`reuse-iters` using sqlite instead of my pieceCompletion storage.
`reuse-iters` maybeUpdateRequest func is not red anymore:
![mem](https://user-images.githubusercontent.com/46885206/167796724-8e8f82ce-997e-481b-b040-aa29f3c8f6c8.png)
![cpu](https://user-images.githubusercontent.com/46885206/167796746-df894b26-e914-4078-a677-feec57af54fd.png)

**Reply by anacrolix** [reply]
Looks way better?

**Reply by AskAlexSharov** [reply]
Yes

**Reply by anacrolix** [reply]
Thank you so much for testing and persisting through this.

**Reply by AskAlexSharov** [reply]
"reuse-iters using sqlite instead of my pieceCompletion storage." - it was bug on my side

**Comment by anacrolix** [comment, marked as answer]
@AskAlexSharov there were massive performance improvements addressing this exact problem in 43c680ec91bcd65f9673ed509495ab0216040c7b. Released in v1.54.1.
